### PR TITLE
bench: initial add memory tracing to jaeger benches

### PR DESCRIPTION
**Pull request by jacobtrombetta**
# Rationale for this change
Initial memory logging has been added to Jaeger benchmark tracing. Memory logging does add execution time to the overall benchmark, so it is placed at `TRACE` level and requires benchmarks to be run with the environment variable `RUST_LOG=trace` to execute.

The logging function does not impact performance when code runs at `DEBUG` level or above.

This benchmark is prior to the change 
image
vs with the memory logging
image

# What changes are included in this PR?
- A `log` module is added to the `utils` crate and the function `log_memory_usage` is implemented.
- Memory logging is added to instrumented sections of code.
- Memory logging is placed at the `TRACE` level, requiring benchmarks to be run as `RUST_LOG=trace` in order to be viewed in the Jaeger UI.

# Are these changes tested?
Yes

### CODE REVIEW

File: crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_commitment_helper_gpu.rs [hunk 1]
```diff
@@ -27,13 +45,18 @@ use tracing::{span, Level};
...
-    skip_all
+    skip_all,
+    fields(total_memory, used_memory)
 )]
 pub(super) fn compute_dynamic_dory_commitments(
     committable_columns: &[CommittableColumn],
     offset: usize,
     setup: &ProverSetup,
 ) -> Vec<DynamicDoryCommitment> {
+    if tracing::level_enabled!(Level::DEBUG) {
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_commitment_helper_gpu.rs:56]: Would be more compact to put this if within `log_memory_usage()`.

File: crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_commitment_helper_gpu.rs [hunk 2]
```diff
@@ -27,13 +45,18 @@ use tracing::{span, Level};
 #[tracing::instrument(
     name = "compute_dynamic_dory_commitments (gpu)",
     level = "debug",
-    skip_all
+    skip_all,
+    fields(total_memory, used_memory)
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_commitment_helper_gpu.rs:49]: Percent used was a nice metric to have.
Can we have `start_` and `end_` versions of these as well?
In other words, put a `log_memory_start_usage()` and `log_memory_end_usage()`.

That way we can see the changed that occurred within a single span just by looking at that span.

File: crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_commitment_helper_gpu.rs (hunk 1 above)
Comment (jacobtrombetta) [reply crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_commitment_helper_gpu.rs]: Agreed. Moved.

File: crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_commitment_helper_gpu.rs (hunk 2 above)
Comment (jacobtrombetta) [reply crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_commitment_helper_gpu.rs]: Updated, but I'm not sure if it fits with what you were thinking?

Comment (JayWhite2357) [reply crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_commitment_helper_gpu.rs]: Does this work? Does it show all 6 things in the span inside jaeger?

I was expecting `fields(available_memory_start, used_memory_start, percentage_memory_used_start, available_memory_end, used_memory_end, percentage_memory_used_end)` or something. But, if this works, I like this better.

Comment (jacobtrombetta) [reply crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_commitment_helper_gpu.rs]: Here is what it looks like with just `fields(available_memory, used_memory, percentage_memory_used)`, from [this commit](https://github.com/spaceandtimelabs/sxt-proof-of-sql/pull/412/commits/5a5c190376f228d2866f7bc87b9e4450eb9c2d34). image

File: crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_commitment_helper_gpu.rs [hunk 3]
```diff
@@ -7,7 +7,28 @@ use crate::base::{commitment::CommittableColumn, if_rayon, slice_ops::slice_cast
...
+        let used_memory = system.used_memory() as f64 / 1024.0;
+        let percentage_memory_used = (used_memory / (used_memory + available_memory)) * 100.0;
+
+        trace!(
+            "{} Available memory: {:.2} MB, Used memory: {:.2} MB, Percentage memory used: {:.2}%",
+            name,
+            available_memory,
+            used_memory,
+            percentage_memory_used
+        );
```
Comment (JayWhite2357) [inline crates/proof-of-sql/src/proof_primitive/dory/dynamic_dory_commitment_helper_gpu.rs:29]: I'm a bit confused. The screenshot looks correct, but the code looks wrong. The `sysinfo` documentation states:
> Returns the amount of available RAM in bytes.

Dividing by 1024 gives KiB, not MiB. So, I'm thinking that the number on the screenshot is KiB, not MB.

But, if that's true, the screenshot implies that there is only 317440.51 KiB = 30 MiB total memory.

**Review by JayWhite2357 — APPROVED** [review]

**Pull request merged**